cpp/src/tests/arrow_compute_test_window.cc

```diff
@@ -0,0 +1,202 @@
+  auto sch = arrow::schema({field("col_int", arrow::int32()), field("col_dou", arrow::float64())});
+  std::vector<std::string> input_data = {
+      "[1, 2, 1]",
+      "[35.612, 37.244, 82.664]"};
+  MakeInputBatch(input_data, sch, &input_batch);
+
+  std::shared_ptr<Field> res = field("window_res", arrow::float64());
+
+  auto f_window = TreeExprBuilder::MakeExpression(TreeExprBuilder::MakeFunction("window", {
+      TreeExprBuilder::MakeFunction("sum",
+          {TreeExprBuilder::MakeField(field("col_dou", arrow::float64()))}, null()),
+      TreeExprBuilder::MakeFunction("partitionSpec",
+          {TreeExprBuilder::MakeField(field("col_int", arrow::int32()))}, null()),
+  }, binary()), res);
+
+  arrow::compute::ExecContext ctx;
+  std::shared_ptr<CodeGenerator> expr;
+  std::vector<std::shared_ptr<arrow::RecordBatch>> out;
+  ASSERT_NOT_OK(
+      CreateCodeGenerator(ctx.memory_pool(), sch, {f_window}, {res}, &expr, true))
+  ASSERT_NOT_OK(expr->evaluate(input_batch, nullptr))
+  ASSERT_NOT_OK(expr->finish(&out))
+
+  std::shared_ptr<arrow::RecordBatch> expected_result;
+  std::vector<std::string> expected_output_data = {
+      "[118.276, 37.244, 118.276]"};
+
+  MakeInputBatch(expected_output_data, arrow::schema({res}), &expected_result);
+  ASSERT_NOT_OK(Equals(*expected_result.get(), *(out.at(0).get())));
+}
+
+TEST(TestArrowComputeWindow, DecimalTest) {
+  std::shared_ptr<arrow::RecordBatch> input_batch;
+  auto sch = arrow::schema({field("col_int", arrow::int32()), field("col_dec", arrow::decimal128(8, 3))});
+  std::vector<std::string> input_data = {
+      "[1, 2, 1]",
+      "[\"35.612\", \"37.244\", \"82.664\"]"};
+  MakeInputBatch(input_data, sch, &input_batch);
+
+  std::shared_ptr<Field> res = field("window_res", arrow::decimal128(8, 3));
+
+  auto f_window = TreeExprBuilder::MakeExpression(TreeExprBuilder::MakeFunction("window", {
+      TreeExprBuilder::MakeFunction("sum",
+                                    {TreeExprBuilder::MakeField(field("col_dec", arrow::decimal128(8, 3)))}, null()),
+      TreeExprBuilder::MakeFunction("partitionSpec",
+                                    {TreeExprBuilder::MakeField(field("col_int", arrow::int32()))}, null()),
+  }, binary()), res);
+
+  arrow::compute::ExecContext ctx;
+  std::shared_ptr<CodeGenerator> expr;
+  std::vector<std::shared_ptr<arrow::RecordBatch>> out;
+  ASSERT_NOT_OK(
+      CreateCodeGenerator(ctx.memory_pool(), sch, {f_window}, {res}, &expr, true))
+  ASSERT_NOT_OK(expr->evaluate(input_batch, nullptr))
+  ASSERT_NOT_OK(expr->finish(&out))
+
+  std::shared_ptr<arrow::RecordBatch> expected_result;
+  std::vector<std::string> expected_output_data = {
+      "[\"118.276\", \"37.244\", \"118.276\"]"};
+
+  MakeInputBatch(expected_output_data, arrow::schema({res}), &expected_result);
+  ASSERT_NOT_OK(Equals(*expected_result.get(), *(out.at(0).get())));
+}
+
+TEST(TestArrowComputeWindow, DecimalAvgTest) {
+  return; // fixme decimal avg not supported?
```

Comment:
spark will cast decimal data to double in hash agg, i think we could follow this in window